File: scripts/import-enemy-sprites.mjs
#!/usr/bin/env node
/**
 * Import enemy sprites as 128x128 atlases and wire them into the manifest.
 *
 * Usage:
 *   node scripts/import-enemy-sprites.mjs /path/to/downloaded/pngs
 *   # or with npm
 *   npm run assets:import:enemies -- /path/to/downloaded/pngs
 *
 * - Sorts the source PNGs by creation time (ctime) to match the desired order
 * - Renames/maps to ids from content/enemies/core-enemies.json
 * - Resizes to 128x128 using nearest-neighbor (crisp)
 * - Saves to assets/generated/pixel/atlas/<enemy_id>.png
 * - Creates TexturePacker-like single-frame meta in assets/generated/pixel/meta/<enemy_id>.json
 * - Updates assets/generated/pixel/manifest.json with atlas + combatant bindings
 *
 * Requires: sharp (npm i -D sharp)
 */
import { promises as fs } from 'node:fs';
import path from 'node:path';
import process from 'node:process';

let sharp;
try {
  // Lazy import to give a better error if missing
  ({ default: sharp } = await import('sharp'));
} catch (err) {
  console.error('[import-enemy-sprites] Missing dependency: sharp');
  console.error('  Install with: npm i -D sharp');
  process.exit(2);
}

const root = process.cwd();
const ASSETS_ROOT = path.join(root, 'assets', 'generated', 'pixel');
const ATLAS_DIR = path.join(ASSETS_ROOT, 'atlas');
const META_DIR = path.join(ASSETS_ROOT, 'meta');
const MANIFEST_PATH = path.join(ASSETS_ROOT, 'manifest.json');
const ENEMIES_JSON = path.join(root, 'content', 'enemies', 'core-enemies.json');
const SIZE = 128;

const ORDERED_IDS = [
  'saqueador_rio',
  'besta_do_cipo',
  'atirador_fantasma',
  'bruxo_gramofone',
  'capataz_alfandega',
  'sentinela_nevoa',
  'jaguar_lunar',
  'capitao_vesper',
  'capanga_vesper',
];

const STYLE_BY_TAG = {
  machine: 'guardian',
  construct: 'guardian',
  cult: 'spitter',
  beast: 'stalker',
};

const ensureDir = async (dirPath) => {
  await fs.mkdir(dirPath, { recursive: true });
};

const exists = async (filePath) => {
  try {
    await fs.access(filePath);
    return true;
  } catch {
    return false;
  }
};

const readJson = async (filePath, fallback) => {
  try {
    const raw = await fs.readFile(filePath, 'utf8');
    return JSON.parse(raw);
  } catch {
    return fallback;
  }
};

const styleForTags = (tags) => {
  for (const tag of Array.isArray(tags) ? tags : []) {
    if (STYLE_BY_TAG[tag]) return STYLE_BY_TAG[tag];
  }
  return 'stalker';
};

const sortPngsByCtime = async (srcDir) => {
  const entries = await fs.readdir(srcDir, { withFileTypes: true });
  const files = entries
    .filter((e) => e.isFile() && e.name.toLowerCase().endsWith('.png'))
    .map((e) => path.join(srcDir, e.name));
  const stats = await Promise.all(files.map((f) => fs.stat(f).then((s) => ({ f, s }))));
  stats.sort((a, b) => a.s.ctimeMs - b.s.ctimeMs);
  return stats.map((it) => it.f);
};

const upsertAtlas = (manifest, atlasId, relPath, width, height) => {
  const atlases = Array.isArray(manifest.atlases) ? manifest.atlases : (manifest.atlases = []);
  const found = atlases.find((a) => a && a.id === atlasId);
  if (found) {
    found.path = relPath;
    found.width = width;
    found.height = height;
    found.domain = 'combat';
    return;
  }
  atlases.push({ id: atlasId, path: relPath, width, height, domain: 'combat' });
};

const upsertCombatantBinding = (manifest, visualKey, atlasId, metaRel, style) => {
  const combatants = Array.isArray(manifest.combatants)
    ? manifest.combatants
    : (manifest.combatants = []);
  const found = combatants.find((c) => c && c.visualKey === visualKey);
  const base = {
    visualKey,
    source: 'texturepacker',
    atlasId,
    metaPath: metaRel,
    width: SIZE,
    height: SIZE,
    anchorX: 0.5,
    anchorY: 1,
    style,
  };
  if (found) {
    Object.assign(found, base);
  } else {
    combatants.push(base);
  }
};

const writeMeta = async (metaPath) => {
  const meta = {
    frames: {
      'idle_0.png': {
        frame: { x: 0, y: 0, w: SIZE, h: SIZE },
        duration: 200,
      },
    },
  };
  await fs.writeFile(metaPath, `${JSON.stringify(meta, null, 2)}\n`, 'utf8');
};

const main = async () => {
  const srcDir = process.argv[2];
  if (!srcDir) {
    console.error('Uso: node scripts/import-enemy-sprites.mjs /caminho/para/pngs');
    process.exit(2);
  }

  await ensureDir(ATLAS_DIR);
  await ensureDir(META_DIR);

  const enemies = await readJson(ENEMIES_JSON, []);
  const indexById = new Map(
    (Array.isArray(enemies) ? enemies : []).map((e) => [e?.id, e]).filter(([k]) => typeof k === 'string'),
  );

  const manifest =
    (await readJson(MANIFEST_PATH, null)) ?? {
      version: 1,
      generatedAt: new Date(0).toISOString(),
      atlases: [],
      combatants: [],
      uiIcons: [],
      diceFaces: [],
      biomeThemes: [],
    };

  const sources = await sortPngsByCtime(srcDir);
  if (sources.length < ORDERED_IDS.length) {
    console.warn(
      `[import-enemy-sprites] Aviso: ${sources.length} arquivos encontrados; ${ORDERED_IDS.length} exigidos.`,
    );
  }

  for (let i = 0; i < ORDERED_IDS.length; i++) {
    const enemyId = ORDERED_IDS[i];
    const src = sources[i];
    if (!src) {
      console.warn(`[import-enemy-sprites] Faltando imagem para ${enemyId}; ignorando.`);
      continue;
    }

    const outPng = path.join(ATLAS_DIR, `${enemyId}.png`);
    const outMeta = path.join(META_DIR, `${enemyId}.json`);
    const atlasId = `enemy.${enemyId}`;
    const relPng = `atlas/${enemyId}.png`;
    const relMeta = `meta/${enemyId}.json`;

    // Resize with nearest neighbor and save as PNG
    await sharp(src).resize(SIZE, SIZE, { kernel: sharp.kernel.nearest, fit: 'fill' }).png().toFile(outPng);

    await writeMeta(outMeta);

    const tags = indexById.get(enemyId)?.tags ?? [];
    const style = styleForTags(tags);

    upsertAtlas(manifest, atlasId, relPng, SIZE, SIZE);
    upsertCombatantBinding(manifest, `enemy:${enemyId}`, atlasId, relMeta, style);

    console.log(`[import-enemy-sprites] OK: ${path.basename(src)} → ${relPng} (visualKey=enemy:${enemyId})`);
  }

  if (!manifest.generatedAt) {
    manifest.generatedAt = new Date(0).toISOString();
  }
  await fs.writeFile(MANIFEST_PATH, `${JSON.stringify(manifest, null, 2)}\n`, 'utf8');
  console.log(`[import-enemy-sprites] Manifest atualizado em ${path.relative(root, MANIFEST_PATH)}`);
};

main().catch((err) => {
  console.error('[import-enemy-sprites] Erro inesperado:', err);
  process.exit(1);
});

File: scripts/import-enemy-sprites.py
#!/usr/bin/env python3
"""
Import enemy sprites as 128x128 atlases and wire them into the manifest.

Usage:
  python3 scripts/import-enemy-sprites.py /path/to/downloaded/pngs

- Sorts the source PNGs by creation time (ctime) to match the desired order
- Renames to the expected ids from content/enemies/core-enemies.json
- Resizes to 128x128 using nearest-neighbor (pixel crisp)
- Saves to assets/generated/pixel/atlas/<enemy_id>.png
- Creates TexturePacker-like single-frame meta in assets/generated/pixel/meta/<enemy_id>.json
- Updates assets/generated/pixel/manifest.json with atlas + combatant bindings

Requires: Pillow (pip install Pillow)
"""
from __future__ import annotations
import json
import os
import sys
from pathlib import Path

try:
    from PIL import Image
except Exception as exc:  # pragma: no cover
    print("[import-enemy-sprites] Pillow not found. Install with: pip install Pillow")
    raise

ROOT = Path(__file__).resolve().parent.parent
ASSETS_ROOT = ROOT / 'assets' / 'generated' / 'pixel'
ATLAS_DIR = ASSETS_ROOT / 'atlas'
META_DIR = ASSETS_ROOT / 'meta'
MANIFEST_PATH = ASSETS_ROOT / 'manifest.json'
ENEMIES_JSON = ROOT / 'content' / 'enemies' / 'core-enemies.json'
SIZE = 128

ORDERED_IDS = [
    'saqueador_rio',
    'besta_do_cipo',
    'atirador_fantasma',
    'bruxo_gramofone',
    'capataz_alfandega',
    'sentinela_nevoa',
    'jaguar_lunar',
    'capitao_vesper',
    'capanga_vesper',
]

STYLE_BY_TAG = {
    'machine': 'guardian',
    'construct': 'guardian',
    'cult': 'spitter',
    'beast': 'stalker',
}


def style_for_tags(tags: list[str]) -> str:
    for tag in tags:
        if tag in STYLE_BY_TAG:
            return STYLE_BY_TAG[tag]
    return 'stalker'


def load_enemies_index() -> dict[str, dict]:
    data = json.loads(ENEMIES_JSON.read_text('utf-8'))
    index = {entry['id']: entry for entry in data if isinstance(entry, dict) and 'id' in entry}
    return index


def read_manifest() -> dict:
    if MANIFEST_PATH.exists():
        return json.loads(MANIFEST_PATH.read_text('utf-8'))
    return {
        'version': 1,
        'generatedAt': '1970-01-01T00:00:00.000Z',
        'atlases': [],
        'combatants': [],
        'uiIcons': [],
        'diceFaces': [],
        'biomeThemes': [],
    }


def ensure_dirs() -> None:
    ATLAS_DIR.mkdir(parents=True, exist_ok=True)
    META_DIR.mkdir(parents=True, exist_ok=True)


def sort_pngs_by_ctime(src: Path) -> list[Path]:
    files = [p for p in src.iterdir() if p.is_file() and p.suffix.lower() == '.png']
    files.sort(key=lambda p: p.stat().st_ctime)
    return files


def upsert_atlas(manifest: dict, atlas_id: str, rel_path: str, width: int, height: int) -> None:
    atlases: list = manifest.get('atlases', [])
    for entry in atlases:
        if isinstance(entry, dict) and entry.get('id') == atlas_id:
            entry['path'] = rel_path
            entry['width'] = width
            entry['height'] = height
            entry['domain'] = 'combat'
            return
    atlases.append({
        'id': atlas_id,
        'path': rel_path,
        'width': width,
        'height': height,
        'domain': 'combat',
    })
    manifest['atlases'] = atlases


def upsert_combatant_binding(manifest: dict, visual_key: str, atlas_id: str, meta_rel: str, style: str) -> None:
    combatants: list = manifest.get('combatants', [])
    for entry in combatants:
        if isinstance(entry, dict) and entry.get('visualKey') == visual_key:
            # Keep any non-procedural binding; update fields for texturepacker binding
            entry.update({
                'visualKey': visual_key,
                'source': 'texturepacker',
                'atlasId': atlas_id,
                'metaPath': meta_rel,
                'width': SIZE,
                'height': SIZE,
                'anchorX': 0.5,
                'anchorY': 1,
                'style': style,
            })
            manifest['combatants'] = combatants
            return
    combatants.append({
        'visualKey': visual_key,
        'source': 'texturepacker',
        'atlasId': atlas_id,
        'metaPath': meta_rel,
        'width': SIZE,
        'height': SIZE,
        'anchorX': 0.5,
        'anchorY': 1,
        'style': style,
    })
    manifest['combatants'] = combatants


def write_meta(meta_path: Path) -> None:
    meta = {
        'frames': {
            'idle_0.png': {
                'frame': { 'x': 0, 'y': 0, 'w': SIZE, 'h': SIZE },
                'duration': 200,
            }
        }
    }
    meta_path.write_text(json.dumps(meta, indent=2) + '\n', encoding='utf-8')


def process_images(src_dir: Path) -> None:
    ensure_dirs()
    enemies_index = load_enemies_index()
    manifest = read_manifest()

    sources = sort_pngs_by_ctime(src_dir)
    if len(sources) < len(ORDERED_IDS):
        print(f"[import-enemy-sprites] Aviso: {len(sources)} arquivos encontrados; {len(ORDERED_IDS)} exigidos.")

    for i, enemy_id in enumerate(ORDERED_IDS):
        if i >= len(sources):
            print(f"[import-enemy-sprites] Faltando imagem para {enemy_id}; ignorando.")
            continue

        src_path = sources[i]
        out_png = ATLAS_DIR / f"{enemy_id}.png"
        out_meta = META_DIR / f"{enemy_id}.json"
        atlas_id = f"enemy.{enemy_id}"
        rel_png = f"atlas/{enemy_id}.png"
        rel_meta = f"meta/{enemy_id}.json"

        # Load → convert → resize → save (nearest)
        img = Image.open(src_path).convert('RGBA')
        img = img.resize((SIZE, SIZE), resample=Image.NEAREST)
        img.save(out_png)

        write_meta(out_meta)

        tags = enemies_index.get(enemy_id, {}).get('tags', [])
        style = style_for_tags(tags)

        upsert_atlas(manifest, atlas_id, rel_png, SIZE, SIZE)
        upsert_combatant_binding(manifest, f"enemy:{enemy_id}", atlas_id, rel_meta, style)

        print(f"[import-enemy-sprites] OK: {src_path.name} → {rel_png} (visualKey=enemy:{enemy_id})")

    manifest['generatedAt'] = manifest.get('generatedAt') or '1970-01-01T00:00:00.000Z'
    MANIFEST_PATH.write_text(json.dumps(manifest, indent=2) + '\n', encoding='utf-8')
    print(f"[import-enemy-sprites] Manifest atualizado em {MANIFEST_PATH.relative_to(ROOT)}")


if __name__ == '__main__':
    if len(sys.argv) < 2:
        print("Uso: python3 scripts/import-enemy-sprites.py /caminho/para/pngs")
        sys.exit(2)
    src_dir = Path(sys.argv[1]).expanduser().resolve()
    if not src_dir.exists():
        print(f"Diretorio nao encontrado: {src_dir}")
        sys.exit(2)
    process_images(src_dir)
